### PR TITLE
feat(SCT-1506): Mash referral final decision

### PR DESCRIPTION
**Pull request by JohnFarrellDev**
## Link to JIRA ticket

[Ticket](https://hackney.atlassian.net/browse/SCT-1506)

### *What is the problem we're trying to solve*

API to take the final decision from the final decision form and update the referral.

### *What changes have we introduced*

Added a FINAL-DECISION stage, applied validation, if valid update referral.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs [hunk 1]
```diff
@@ -85,7 +85,7 @@ public void WhenUpdateTypeIsInitialDecisionRequiresUrgentContactMustBeProvided()
         [Test]
         public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustNotBeNull()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs:86]: This test reads initial decision in the description but we pass "FINAL-DECISION" as the update type within the test.

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs [hunk 2]
```diff
@@ -96,7 +96,62 @@ public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustNotBeNull()
         [Test]
         public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustHaveALength()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs:97]: Same as above.

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs [hunk 3]
```diff
@@ -96,7 +96,62 @@ public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustNotBeNull()
...
+            var request = TestHelpers.CreateUpdateMashReferral(updateType: "FINAL-DECISION");
+            request.ReferralCategory = "";
+
+            var validationResult = _validator.Validate(request);
+
+            validationResult.ToString().Should().Be("Must provide a referral category");
+        }
+
+        [Test]
+        public void WhenUpdateTypeIsInitialDecisionADecisionMustNotBeNull()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs:108]: Same as above.

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs [hunk 4]
```diff
@@ -96,7 +96,62 @@ public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustNotBeNull()
...
+            var request = TestHelpers.CreateUpdateMashReferral(updateType: "FINAL-DECISION");
+            request.Decision = null;
+
+            var validationResult = _validator.Validate(request);
+
+            validationResult.ToString().Should().Be("Must provide a decision");
+        }
+
+        [Test]
+        public void WhenUpdateTypeIsInitialDecisionDecisionMustHaveALength()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs:119]: Same as above

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs [hunk 5]
```diff
@@ -96,7 +96,62 @@ public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustNotBeNull()
...
+            var request = TestHelpers.CreateUpdateMashReferral(updateType: "FINAL-DECISION");
+            request.RequiresUrgentContact = null;
+
+            var validationResult = _validator.Validate(request);
+
+            validationResult.ToString().Should().Be("Must provide if urgent contact is required");
+        }
+
+        [Test]
+        public void WhenUpdateTypeIsFinalDecisionReferralCategoryMustNotBeNull()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs:141]: This looks like it shares the same setup as the test on line 86. Not sure if we've duplicated the test (it has a different test name but same setup as this)or maybe we've passed in the wrong updateType for that test on line 86?

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs [hunk 6]
```diff
@@ -96,7 +96,62 @@ public void WhenUpdateTypeIsInitialDecisionReferralCategoryMustNotBeNull()
...
+            var request = TestHelpers.CreateUpdateMashReferral(updateType: "FINAL-DECISION");
+            request.ReferralCategory = null;
+
+            var validationResult = _validator.Validate(request);
+
+            validationResult.ToString().Should().Be("Must provide a referral category");
+        }
+
+        [Test]
+        public void WhenUpdateTypeIsFinalDecisionReferralCategoryMustHaveALength()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateMashReferralTests.cs:152]: Same as above. (Looks similar to the test on line 97.

**Review by Renny4Real — COMMENTED** [review]
🎉 Just a few comments. 🙏🏾

**Comment by JohnFarrellDev** [comment]
@Renny4Real thank you for catching those naming mis-matches

**Pull request merged**